tests/utils/test_schemapi.py

```diff
@@ -948,3 +952,37 @@ def test_to_dict_expand_mark_spec():
     chart = alt.Chart().mark_bar()
     assert chart.to_dict()["mark"] == {"type": "bar"}
     assert chart.mark == "bar"
+
+
+@pytest.mark.parametrize(
+    "expected",
+    [list("cdfabe"), [0, 3, 4, 5, 8]],
+)
+@pytest.mark.parametrize(
+    "tp",
+    [
+        tuple,
+        list,
+        deque,
+        pl.Series,
+        pd.Series,
+        pd.Index,
+        pd.Categorical,
+        pd.CategoricalIndex,
+        np.array,
+    ],
+)
```

Comment:
Should this include `set`?

Reply:
> Should this include `set`?

It could, but would require adapting the test as bit more to account for `set` not being ordered:

```py
    def test_to_dict_iterables(tp, expected) -> None:
        x_dict = alt.X("x:N", sort=tp(expected)).to_dict()
        actual = x_dict["sort"]  # type: ignore
>       assert actual == expected
E       AssertionError: assert ['e', 'b', 'c', 'a', 'f', 'd'] == ['c', 'd', 'f', 'a', 'b', 'e']
E
E         At index 0 diff: 'e' != 'c'
E         Use -v to get more diff

tests\utils\test_schemapi.py:979: AssertionError
================================================================================================ short test summary info ================================================================================================ 
FAILED tests/utils/test_schemapi.py::test_to_dict_iterables[set-expected0] - AssertionError: assert ['e', 'b', 'c', 'a', 'f', 'd'] == ['c', 'd', 'f', 'a', 'b', 'e']
============================================================================================= 1 failed, 19 passed in 18.88s ============================================================================================= 

```

Related to this, all the types I chose are `Sequence` except the `numpy.array` as mentioned here https://github.com/vega/altair/issues/2877

So while internally we are using `Iterable`, currently all of the public API uses the annotation `Sequence` and `set` is the former but not the latter.

Reply:
Great point, maybe we should strictly adhere to passing only ordered sequences/arrays and disallow passing unordered sets? To me that seems closer to the intended use case in both the examples motivating this PR: the user wants to specify both the elements and in which order they occur. It might be confusing if they pass a set without realizing that it is unordered and altair does not show an error but instead the elements show up in the wrong order. What do you think?

Reply:
I think I answered more of this than I planned to in https://github.com/vega/altair/pull/3501#issuecomment-2269295210 so I'll just respond to the bits I don't think were covered there

> It might be confusing if they pass a set without realizing that it is unordered and altair does not show an error but instead the elements show up in the wrong order. What do you think?

I totally get the intention, but if a user wants to use the standard library incorrectly - who are we to stop them?
The **second sentence** in the [python tutorial on sets](https://docs.python.org/3/tutorial/datastructures.html#sets) states they are unordered.

---

Additionally, passing a `set` will *fail type checking* for `Sequence`, which is what I meant when I said:

> So while internally we are using `Iterable`, currently all of the public API uses the annotation `Sequence` and `set` is the former but not the latter.

```py
>>> mypy tests
error: Argument "sort" to "X" has incompatible type "set[str]"; expected
"dict[Any, Any] | SchemaBase | Sequence[str] | Sequence[bool] | Sequence[float] | Sequence[dict[Any, Any] | SchemaBase] | Literal['ascending', 'descending', 'x', 'y', 'color', 'fill', 'stroke', 'strokeWidth', 'size', 'shape', 'fillOpacity', 'strokeOpacity', 'opacity', 'text', '-x', '-y', '-color', '-fill', '-stroke', '-strokeWidth', '-size', '-shape', '-fillOpacity', '-strokeOpacity', '-opacity', '-text'] | UndefinedType | None"
 [arg-type]
        attempt_set = alt.X("fieldname", sort={"1", "2", "3"})
                                              ^~~~~~~~~~~~~~~ 
Found 1 error in 1 file (checked 325 source files)
```

Reply:
Thanks for adding the additional tests accounting for more cases than `sort`. I agree that it is not Altair's responsibility to make sure users are using other libraries correctly. At the same time I think hints of correct usage with Altair can be helpful but as you said we already have that via the `Sequence` typing hint, which should be sufficient so let's leave this as is.